spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/SparkParquetFileMergeRunner.java

```diff
@@ -0,0 +1,404 @@
+ * without full deserialization, which is significantly faster than the standard Spark rewrite
+ * approach for Parquet files.
+ *
+ * <p>The decision to use this runner vs. SparkBinPackFileRewriteRunner is controlled by the
+ * configuration option {@code use-parquet-file-merger}.
+ */
+public class SparkParquetFileMergeRunner extends SparkBinPackFileRewriteRunner {
+  private static final Logger LOG = LoggerFactory.getLogger(SparkParquetFileMergeRunner.class);
+  private final FileRewriteCoordinator coordinator = FileRewriteCoordinator.get();
+
+  public SparkParquetFileMergeRunner(SparkSession spark, Table table) {
+    super(spark, table);
+  }
+
+  @Override
+  public String description() {
+    return "PARQUET-MERGE";
+  }
+
+  @Override
+  protected void doRewrite(String groupId, RewriteFileGroup group) {
+    // Early validation: check if requirements are met
+    if (!canMerge(group)) {
+      LOG.info(
+          "Row-group merge requirements not met for group {}. Using standard Spark rewrite.",
+          groupId);
+      super.doRewrite(groupId, group);
+      return;
+    }
+
+    // Requirements met - attempt row-group level merge
+    try {
+      LOG.info(
+          "Merging {} Parquet files using row-group level merge (group: {})",
+          group.rewrittenFiles().size(),
+          groupId);
+      mergeParquetFilesDistributed(groupId, group);
+    } catch (Exception e) {
+      LOG.info(
+          "Row-group merge failed for group {}, falling back to standard Spark rewrite: {}",
+          groupId,
+          e.getMessage(),
+          e);
+      // Fallback to standard rewrite
+      super.doRewrite(groupId, group);
+    }
+  }
+
+  /**
+   * Validates if the file group can use row-group level merging.
+   *
+   * <p>Requirements checked here:
+   *
+   * <ul>
+   *   <li>Group must expect exactly 1 output file (ParquetFileMerger cannot split into multiple
+   *       files)
+   *   <li>Table must not have a sort order (no sorting or z-ordering)
+   *   <li>Files must not have delete files or delete vectors
+   *   <li>Partition spec must not be changing (row-group merge cannot repartition data)
+   *   <li>Must not be splitting large files (row-group merge is optimized for merging, not
+   *       splitting)
+   * </ul>
+   *
+   * <p>Additional requirements checked by ParquetFileMerger.canMerge:
+   *
+   * <ul>
+   *   <li>All files must be valid Parquet format
+   *   <li>Files must have compatible schemas
+   *   <li>Files must not be encrypted
+   *   <li>If a physical _row_id column exists, all values must be non-null
+   * </ul>
+   *
+   * <p>To ensure groups produce single output files, configure: {@code
+   * rewrite-data-files.max-file-group-size-bytes = rewrite-data-files.target-file-size-bytes}
+   *
+   * @param group the file group to check
+   * @return true if row-group merging can be used, false otherwise
+   */
+  @VisibleForTesting
+  boolean canMerge(RewriteFileGroup group) {
```

Comment:
Talked to @Guosmilesmile who is working on enabling this feature for the Flink compaction, and he mentioned that he needs to implement these exact checks in Flink as well, so it might be good idea to move this to a common place, like `ParquetFileMerger` immediately

Reply:
What if we would call this method `MessageType mergedSchema(RewriteFileGroup group)` and return the merged file schema, and maybe null if it can't be merged? This way we could run this on the Driver (open every file once there), then put the merged schema to the `MergeTaskInfo` and send it to the Executors. This way the executors don't have to open a file to get the schema again, and every Executor will open every file only once.

Sorry for the back and forth, but I understand the constraints of the Spark solution better now.